File: 享元模式/享元模式/ViewController.m
//
//  ViewController.m
//  享元模式
//
//  Created by qrh on 2019/7/29.
//  Copyright © 2019 zdq. All rights reserved.
//

#import "ViewController.h"
#import "TicketFactory.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
//    享元模式，用于缓存模式
    TicketFactory * factory = [[TicketFactory alloc] init];
    id<TrainTicketProtocol> ticket = [factory ticketFrom:@"杭州" to:@"北京"];
    [ticket showTicketInfo:@"二等座"];
}


@end

File: 备忘录模式/备忘录模式/ViewController.swift
//
//  ViewController.swift
//  备忘录模式
//
//  Created by qrh on 2019/7/30.
//  Copyright © 2019 zdq. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        
//        开始游戏
        let originator = JDQSOriginator()
        originator.play()
        originator.play()

//        创建备忘录,存档
        let caretaker = JDQSCaretaker()
        let memo = originator.createMemo();
        caretaker.write(memo: memo)

//        c退出游戏
        originator.exit()
        
//        恢复游戏
        let restoreOrg = JDQSOriginator()
        restoreOrg.restore(memo: memo)
    
        
    }


}

